If module is missing a docstring show readme if available

IanButterworth opened the pull request:
(re)enables printing of readme's, if the module is missing a docstring. (supposedly this used to happen in some 0.x versions)

Recursively climbs up the path to the root package directory looking for readmes, in case a specific nested module has a dedicated readme at that directory level.

```
help?> LLVM_full_jll
search: LLVM_full_jll

  No docstring found for module LLVM_full_jll.

  Displaying the contents of /Users/ian/.julia/packages/LLVM_full_jll/YJHoG/README.md:

  LLVM_full_jll.jl (v11.0.0+9)
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡

  This is an autogenerated package constructed using BinaryBuilder.jl (https://github.com/JuliaPackaging/BinaryBuilder.jl). The originating
  build_tarballs.jl
  (https://github.com/JuliaPackaging/Yggdrasil/blob/19767c85345c2f81b7451344b87cc295dc5d8ac3/L/LLVM/LLVM_full@11.0.0/build_tarballs.jl) script can be
  found on Yggdrasil (https://github.com/JuliaPackaging/Yggdrasil/), the community build tree. If you have any issue, please report it to the Yggdrasil
  bug tracker (https://github.com/JuliaPackaging/Yggdrasil/issues).

  For more details about JLL packages and how to use them, see BinaryBuilder.jl documentation
  (https://juliapackaging.github.io/BinaryBuilder.jl/dev/jll/).

  Sources
  =========

  The tarballs for LLVM_full_jll.jl have been built from these sources:

    •  git repository: https://github.com/llvm/llvm-project.git (revision: 176249bd6732a8044d457092ed932768724a6f06)

    •  files in directory, relative to originating build_tarballs.jl: ./bundled
       (https://github.com/JuliaPackaging/Yggdrasil/tree/19767c85345c2f81b7451344b87cc295dc5d8ac3/L/LLVM/LLVM_full@11.0.0/bundled)

  Platforms
  ===========

  LLVM_full_jll.jl is available for the following platforms:

    •  macOS aarch64 (aarch64-apple-darwin)

    •  Linux aarch64 {cxxstring_abi=cxx03, libc=glibc} (aarch64-linux-gnu-cxx03)

    •  Linux aarch64 {cxxstring_abi=cxx11, libc=glibc} (aarch64-linux-gnu-cxx11)

    •  Linux aarch64 {cxxstring_abi=cxx03, libc=musl} (aarch64-linux-musl-cxx03)

    •  Linux aarch64 {cxxstring_abi=cxx11, libc=musl} (aarch64-linux-musl-cxx11)

    •  Linux armv6l {call_abi=eabihf, cxxstring_abi=cxx03, libc=glibc} (armv6l-linux-gnueabihf-cxx03)

    •  Linux armv6l {call_abi=eabihf, cxxstring_abi=cxx11, libc=glibc} (armv6l-linux-gnueabihf-cxx11)

    •  Linux armv6l {call_abi=eabihf, cxxstring_abi=cxx03, libc=musl} (armv6l-linux-musleabihf-cxx03)

    •  Linux armv6l {call_abi=eabihf, cxxstring_abi=cxx11, libc=musl} (armv6l-linux-musleabihf-cxx11)

    •  Linux armv7l {call_abi=eabihf, cxxstring_abi=cxx03, libc=glibc} (armv7l-linux-gnueabihf-cxx03)

    •  Linux armv7l {call_abi=eabihf, cxxstring_abi=cxx11, libc=glibc} (armv7l-linux-gnueabihf-cxx11)

    •  Linux armv7l {call_abi=eabihf, cxxstring_abi=cxx03, libc=musl} (armv7l-linux-musleabihf-cxx03)

    •  Linux armv7l {call_abi=eabihf, cxxstring_abi=cxx11, libc=musl} (armv7l-linux-musleabihf-cxx11)

    •  Linux i686 {cxxstring_abi=cxx03, libc=glibc} (i686-linux-gnu-cxx03)

    •  Linux i686 {cxxstring_abi=cxx11, libc=glibc} (i686-linux-gnu-cxx11)

    •  Linux i686 {cxxstring_abi=cxx03, libc=musl} (i686-linux-musl-cxx03)

    •  Linux i686 {cxxstring_abi=cxx11, libc=musl} (i686-linux-musl-cxx11)

    •  Windows i686 {cxxstring_abi=cxx03} (i686-w64-mingw32-cxx03)

    •  Windows i686 {cxxstring_abi=cxx11} (i686-w64-mingw32-cxx11)

    •  Linux powerpc64le {cxxstring_abi=cxx03, libc=glibc} (powerpc64le-linux-gnu-cxx03)

    •  Linux powerpc64le {cxxstring_abi=cxx11, libc=glibc} (powerpc64le-linux-gnu-cxx11)

    •  macOS x86_64 (x86_64-apple-darwin)

    •  Linux x86_64 {cxxstring_abi=cxx03, libc=glibc} (x86_64-linux-gnu-cxx03)

    •  Linux x86_64 {cxxstring_abi=cxx11, libc=glibc} (x86_64-linux-gnu-cxx11)

    •  Linux x86_64 {cxxstring_abi=cxx03, libc=musl} (x86_64-linux-musl-cxx03)

    •  Linux x86_64 {cxxstring_abi=cxx11, libc=musl} (x86_64-linux-musl-cxx11)

    •  FreeBSD x86_64 (x86_64-unknown-freebsd)

    •  Windows x86_64 {cxxstring_abi=cxx03} (x86_64-w64-mingw32-cxx03)

    •  Windows x86_64 {cxxstring_abi=cxx11} (x86_64-w64-mingw32-cxx11)

  Products
  ==========

  The code bindings within this package are autogenerated from the following Products:

    •  ExecutableProduct: clang

    •  LibraryProduct: libclang

    •  LibraryProduct: libllvm

    •  LibraryProduct: liblto

    •  ExecutableProduct: llc

    •  ExecutableProduct: llvm_config

    •  ExecutableProduct: llvm_mca

    •  ExecutableProduct: opt
```

IanButterworth committed:
print available readme if no docstring for module

DilumAluthge commented:
It would be nice to open either an editor or a pager if the README is long.

DilumAluthge commented:
Or, at the very least, show only up to the the first N characters of the README, for a suitable choice of N.

IanButterworth committed:
tweak search & print

IanButterworth committed:
handle pathless modules

ianshmean commented:
> It would be nice to open either an editor or a pager if the README is long.

I think figuring out opening another editor/viewer would be hard to get the UX right across all the different front ends

> Or, at the very least, show only up to the the first N characters of the README, for a suitable choice of N.

Is the annoyance of having too much text print greater than the possible annoyance of the text being cut off just before the bit you need? Perhaps a line limit makes sense, but I'd make it quite large. Large enough to show the entire example above, at least

DilumAluthge commented:
> Perhaps a line limit makes sense, but I'd make it quite large. Large enough to show the entire example above, at least

Yeah you can definitely make it really large.

DilumAluthge commented:
> I think figuring out opening another editor/viewer would be hard to get the UX right across all the different front ends

Could we re-use the functionally of `@edit`.

IanButterworth committed:
limit readme print to first 200 lines

oxinabox commented:
> It would be nice to open either an editor or a pager if the README is long.

Lets leave this for another discussion.
This same problem exists for docstrings in general.
So that solution needs to be broader than  for this PR.
Its a bigger discussion and adding such functionality in this PR would make this PR be doing two things, which would make it a bad PR.


For this PR, I think we should just show the whole thing.
A module docstring that is too long to read is no worse than no docstring at all.
So this PR is a net win already.

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -243,10 +243,12 @@ end
+        !isa(binding_res, Module) && println(io, "No documentation found.\n")
+        summarize(io, binding_res, binding)
     else
+        println(io, "No documentation found.\n")
```
Extra blank lines don't semm useful here
```suggestion
        println(io, "No documentation found.")
```

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -243,10 +243,12 @@ end
     if defined(binding)
-        summarize(io, resolve(binding), binding)
+        binding_res = resolve(binding)
+        !isa(binding_res, Module) && println(io, "No documentation found.\n")
```
Extra blank lines don't semm useful here
```suggestion
        !isa(binding_res, Module) && println(io, "No documentation found.")
```

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -299,8 +301,33 @@ function summarize(io::IO, T::DataType, binding::Binding)
+    path = dirname(pathof(m))
+    top_path = pkgdir(m)
+    while true
+        for readme in ["README.md", "readme.md", "Readme.md", "ReadMe.md"]
+            readme_path = joinpath(path, readme)
+            if isfile(readme_path)
+                return readme_path
+            end
+        end
```
Is there a reason to do this rather than:
```
for file in readdir(path; join=true, sort=true)
    isfile(file) && basename(lowerscase(file)) == "readme.md" || continue
    return file
end
```

I don't think performance matters here, but missing out on a file called `ReAdMe.Md` would be bad.
Or are such file not recognized by github anyway?

fredrikekre replied:
`README` (without the `.md`) is common aswell. I don't think I have seen the other variants, just `README.md` and `README`.

mcabbott commented:
Could this hook into the `# Extended help` story with `??` perhaps? 

While docstrings are written with the intention of being used like this, some readme files are pretty long, written assuming you have a browser.

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -299,8 +301,33 @@ function summarize(io::IO, T::DataType, binding::Binding)
+function summarize(io::IO, m::Module, binding::Binding; nlines::Int = 200)
+    readme_path = find_readme(m)
+    if !isnothing(readme_path)
+        println(io, "No docstring found for module `", m, "`.\n")
```
This white space especially is i think excessive.
Because the next sentence: that says "Dispaying readme..." is a direct continuation from this.
```suggestion
        println(io, "No docstring found for module `$m`.")
```

Also is there a reason for not using interpolation?

ianshmean replied:
just from copying from original code. Updated

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -299,8 +301,33 @@ function summarize(io::IO, T::DataType, binding::Binding)
+    if !isnothing(readme_path)
+        println(io, "No docstring found for module `", m, "`.\n")
+        readme_lines = readlines(readme_path)
+        println(io, "Displaying the contents of `$(readme_path)`:\n\n")
```
1 blank line is enough
```suggestion
        println(io, "Displaying the contents of `$(readme_path)`:\n")
```

ianshmean replied:
I don't understand why, but a simple println doesn't actually line break in this display. It requires an extra `\n` hence all the extra `\n`'s (most of which I just took from the original code when moving things around)

ianshmean replied:
Turns out to be a bug where there's no way to force a single line break in markdown https://github.com/JuliaLang/julia/issues/16004#issuecomment-754904065
Working around it by printing a single line (which is tidier anyway)

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -299,8 +301,33 @@ function summarize(io::IO, T::DataType, binding::Binding)
+    readme_path = find_readme(m)
+    if !isnothing(readme_path)
+        println(io, "No docstring found for module `", m, "`.\n")
+        readme_lines = readlines(readme_path)
```
Why are we defining this up here, and now where it is used?
I am guessing because if it is empy we don't want to display anything?
```suggestion
readme_lines = readlines(readme_path)
isempty(readme_lines) && return  # don't say we are going to print empty file
```
May want to print something to that effect.

I would suggest using `eachline` instead of `readlines` since no need to read whole thing if only printing part.
but `eachline` is stateful, and checking if it `isempty` would consume the first line.

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -299,8 +301,33 @@ function summarize(io::IO, T::DataType, binding::Binding)
+        println(io, "No docstring found for module `", m, "`.\n")
+        readme_lines = readlines(readme_path)
+        println(io, "Displaying the contents of `$(readme_path)`:\n\n")
+        [println(io, line) for line in first(readme_lines, nlines)]
```
This seems a weird place to use a list comprehennsion.
I could understand using a broadcast because it is short, but a list comprehension just seems odd.

```suggestion
        println.(io, first(readme_lines, nlines))
```
or
```suggestion
        println(io, join(first(readme_lines, nlines), "\n")))
```
or probably best:
```suggestion
        for line in line in first(readme_lines, nlines)
            println(io, line)
        end
```

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -299,8 +301,33 @@ function summarize(io::IO, T::DataType, binding::Binding)
+        readme_lines = readlines(readme_path)
+        println(io, "Displaying the contents of `$(readme_path)`:\n\n")
+        [println(io, line) for line in first(readme_lines, nlines)]
+        length(readme_lines) > nlines && println(io, "[output truncated to first $nlines lines]")
```
Here i think therre should be whitespace? to be more clear that it is not part of the document
```suggestion
        length(readme_lines) > nlines && println(io, "\n[output truncated to first $nlines lines]")
```

oxinabox commented on stdlib/REPL/src/docview.jl:
```diff
@@ -299,8 +301,33 @@ function summarize(io::IO, T::DataType, binding::Binding)
+        [println(io, line) for line in first(readme_lines, nlines)]
+        length(readme_lines) > nlines && println(io, "[output truncated to first $nlines lines]")
+    else
+        println(io, "No docstring or readme found for module `", m, "`.\n")
```
```suggestion
        println(io, "No docstring or readme file found for module `", m, "`.\n")
```

oxinabox reviewed (commented):
This is great stuff

IanButterworth committed:
review changes

mgkuhn commented:
I would rather like to see instead some mechanism to nudge package authors into providing a proper introductory docstring for each module, e.g. by running a test for the presence of such docstrings at package registration, or `Documenter.jl` issuing a warning about the absence of at least one module docstring in the documentation. This patch might delay, rather than encourage, the more ubiquitous presence of module docstrings.

README.md files could focus on information for people who have not yet downloaded and installed the module, and therefore their content would be somewhat more orthogonal to module docstrings.

ianshmean commented:
> mechanism to nudge package authors into providing a proper introductory docstring for each module, e.g. by running a test for the presence of such docstrings at package registration, or Documenter.jl  issuing a warning about the absence of at least one module docstring in the documentation

Yeah, module docstrings are best. I don't think it's an either/or though, implementing what you suggest as well as this PR could be a fair way forward

IanButterworth committed:
print message on single line

ianshmean commented:
First info now prints on a single line
```
help?> LLVM_full_jll
search: LLVM_full_jll

  No docstring found for module LLVM_full_jll. Displaying the contents of /Users/ian/.julia/packages/LLVM_full_jll/YJHoG/README.md:

  LLVM_full_jll.jl (v11.0.0+9)
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡
```

aviks commented:
How would this behave for packages added to custom sysimgs?

IanButterworth committed:
handle modules in sysimage that are missing src files

ianshmean commented:
Good point. I added handling if `pathof(m::Module)` returns a path that doesn't exist

ianshmean commented:
@StefanKarpinski Just a bump for review 🙏🏻

IanButterworth committed:
Merge branch 'master' into ib/module_docstring_from_readme

IanButterworth committed:
Merge branch 'master' into ib/module_docstring_from_readme

IanButterworth commented:
Updated given exported names are now shown on master if no module docstring exists:
```
help?> DifferentialEquations
search: DifferentialEquations

  No docstring found for module DifferentialEquations.

  Exported names
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡

  @derivatives, @ode_def, @ode_def_all, @ode_def_bare, @parameters, @register, @variables, AB3, AB4, AB5, ABDF2, ABM32, ABM43, ABM54, AN5, AbstractAnalyticalProblem, AbstractDiffEqArray,
  AbstractMultiScaleArray, AbstractMultiScaleArrayHead, AbstractMultiScaleArrayLeaf, AbstractVectorOfArray, AffineDiffEqOperator, AitkenNeville, AnalyticalProblem, Anas5, ArrayPartition,
  AutoAbstol, AutoDP5, AutoSOSRA2, AutoSOSRI2, AutoSwitch, AutoTsit5, AutoVern6, AutoVern7, AutoVern8, AutoVern9, BAOAB, BS3, BS5, BVProblem, BipartiteGraph, BlackScholesProblem,
  BoundaryValueDiffEq, BoxWedgeTail, BoxWedgeTail!, BracketData, BrownFullBasicInit, BrownianBridge, BrownianBridge!, CFNLIRK3, CFRLDDRK64, CG2, CG3, CKLLSRK43_2, CKLLSRK54_3C,
  ... (manually truncated)

  Displaying contents of readme found at /Users/ian/.julia/packages/DifferentialEquations/HSWeG/README.md
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡

  DifferentialEquations.jl
  ≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡

  (Image: Join the chat at https://gitter.im/JuliaDiffEq/Lobby) (https://gitter.im/JuliaDiffEq/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) (Image:
  Build Status) (https://github.com/SciML/DifferentialEquations.jl/actions?query=workflow%3ACI) (Image: Stable) (http://diffeq.sciml.ai/stable/) (Image: Dev) (http://diffeq.sciml.ai/dev/)
  (Image: DOI) (https://zenodo.org/badge/latestdoi/58516043)

  This is a suite for numerically solving differential equations written in Julia and available for use in Julia, Python, and R. The purpose of this package is to supply efficient Julia
  implementations of solvers for various differential equations. Equations within the realm of this package include:

    •  Discrete equations (function maps, discrete stochastic (Gillespie/Markov) simulations)

    •  Ordinary differential equations (ODEs)

    •  Split and Partitioned ODEs (Symplectic integrators, IMEX Methods)

    •  Stochastic ordinary differential equations (SODEs or SDEs)

    •  Stochastic differential-algebraic equations (SDAEs)

    •  Random differential equations (RODEs or RDEs)

    •  Differential algebraic equations (DAEs)

    •  Delay differential equations (DDEs)

    •  Neutral, retarded, and algebraic delay differential equations (NDDEs, RDDEs, and DDAEs)

    •  Stochastic delay differential equations (SDDEs)

    •  Experimental support for stochastic neutral, retarded, and algebraic delay differential equations (SNDDEs, SRDDEs, and SDDAEs)

    •  Mixed discrete and continuous equations (Hybrid Equations, Jump Diffusions)

    •  (Stochastic) partial differential equations ((S)PDEs) (with both finite difference and finite element methods)
... (manually truncated)
```

compared to 1.6.0
```
help?> DifferentialEquations
search: DifferentialEquations

  No documentation found.

  No docstring found for module DifferentialEquations.
```

IanButterworth committed:
tweak formatting

IanButterworth committed:
fix docstring test
